src/main/scala/sigmastate/eval/TreeBuilding.scala

```diff
@@ -219,29 +242,59 @@ trait TreeBuilding extends RuntimeCosting { IR: Evaluation =>
     }
   }
 
-  def processAstGraph(mainG: PGraph, env: DefEnv, subG: AstGraph, defId: Int): SValue = {
+  private def processAstGraph(mainG: PGraph,
+                              env: DefEnv,
+                              subG: AstGraph,
+                              defId: Int,
+                              constantsProcessing: Option[ConstantsProcessing]): SValue = {
     val valdefs = new ArrayBuffer[ValDef]
     var curId = defId
     var curEnv = env
     for (TableEntry(s, d) <- subG.schedule) {
       if (mainG.hasManyUsagesGlobal(s) && IsContextProperty.unapply(d).isEmpty && IsInternalDef.unapply(d).isEmpty) {
-        val rhs = buildValue(mainG, curEnv, s, curId)
+        val rhs = buildValue(mainG, curEnv, s, curId, constantsProcessing)
         curId += 1
         val vd = ValDef(curId, Seq(), rhs)
         curEnv = curEnv + (s -> (curId, elemToSType(s.elem)))  // assign valId to s, so it can be use in ValUse
         valdefs += vd
       }
     }
     val Seq(root) = subG.roots
-    val rhs = buildValue(mainG, curEnv, root, curId)
+    val rhs = buildValue(mainG, curEnv, root, curId, constantsProcessing)
     val res = if (valdefs.nonEmpty) BlockValue(valdefs.toIndexedSeq, rhs) else rhs
     res
   }
 
-  def buildTree[T <: SType](f: Rep[Context => Any]): Value[T] = {
+  sealed trait ConstantsProcessing
+
+  class ExtractConstants() extends ConstantsProcessing {
+
+    private val store: ArrayBuffer[Constant[SType]] = new ArrayBuffer[Constant[SType]]
+
+    def extract(c: ConstantNode[SType])(implicit builder: SigmaBuilder): sigmastate.Values.ConstantPlaceholder[SType] =
+      this.synchronized {
```

Comment:
@greenhat does it make sense to have this thread-safe paradise island in an ocean of non-thread-safe code?

Reply:
@ergomorphic Nope. Old reflexes. :)